include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h

```diff
@@ -152,6 +152,24 @@ __pattern_transform_scan_base_impl(_ExecutionPolicy&& __exec, _Iterator1 __first
     }
 }
 
+template <sycl::access::mode _Mode1, sycl::access::mode _Mode2, typename _T, typename _Allocator>
+bool
+__check_equal_iterators(sycl_iterator<_Mode1, _T, _Allocator> __it1, sycl_iterator<_Mode2, _T, _Allocator> __it2)
+{
+    if (__it1 != __it2)
+        return false;
```

Comment:
As @akukanov  mentioned offline, it isn't valid to compare these unless we know these are iterators on the same buffer.  At the very least, they should be moved after the check for the buffers.  

However, with the knowledge that a `(sub_buffer == parent_buffer)` is `true`, I don't believe this check is even known to be valid then.  The `idx` within `sycl_iterator` that is checked here is layered on top of the inherent offset for a sub_buffer.  Without a way to check the sub_buffer's offset, if either iterator's buffer is a sub_buffer I think we have to merely return the result of the buffer equality check (and take the performance hit).

If both buffers are not a sub_buffer, and the buffers are equal, then we can check the equality of the iterators like this.

It also may be worth renaming this function, as we are warping the meaning to basically be:

`__check_if_iterator_equality_is_possible()`

True: these iterators may be the same
False: we know these iterators are different

Reply:
> it isn't valid to compare these unless we know these are iterators on the same buffer

Note also that the same concern applies to the original implementation of this function. It is in general not valid, I think, to compare two iterators just on the base that they are of the same type.